builtin/provisioners/chef/resource_provisioner.go

```diff
@@ -0,0 +1,413 @@
+	defaultEnv     = "_default"
+	firstBoot      = "first-boot.json"
+	logfileDir     = "logfiles"
+	linuxConfDir   = "/etc/chef"
+	validationKey  = "validation.pem"
+	windowsConfDir = "C:/chef"
+)
+
+const clientConf = `
+log_location            STDOUT
+chef_server_url         "{{ .ServerURL }}"
+validation_client_name  "{{ .ValidationClientName }}"
+node_name               "{{ .NodeName }}"
+
+{{ if .HTTPProxy }}
+http_proxy          "{{ .HTTPProxy }}"
+ENV['http_proxy'] = "{{ .HTTPProxy }}"
+ENV['HTTP_PROXY'] = "{{ .HTTPProxy }}"
+{{ end }}
+
+{{ if .HTTPSProxy }}
+https_proxy          "{{ .HTTPSProxy }}"
+ENV['https_proxy'] = "{{ .HTTPSProxy }}"
+ENV['HTTPS_PROXY'] = "{{ .HTTPSProxy }}"
+{{ end }}
+
+{{ if .NOProxy }}no_proxy "{{ join .NOProxy "," }}"{{ end }}
+{{ if .SSLVerifyMode }}ssl_verify_mode {{ .SSLVerifyMode }}{{ end }}
+`
+
+// Provisioner represents a specificly configured chef provisioner
+type Provisioner struct {
+	Attributes           interface{} `mapstructure:"attributes"`
+	Environment          string      `mapstructure:"environment"`
+	LogToFile            bool        `mapstructure:"log_to_file"`
+	HTTPProxy            string      `mapstructure:"http_proxy"`
+	HTTPSProxy           string      `mapstructure:"https_proxy"`
+	NOProxy              []string    `mapstructure:"no_proxy"`
+	NodeName             string      `mapstructure:"node_name"`
+	PreventSudo          bool        `mapstructure:"prevent_sudo"`
+	RunList              []string    `mapstructure:"run_list"`
+	ServerURL            string      `mapstructure:"server_url"`
+	SkipInstall          bool        `mapstructure:"skip_install"`
+	SSLVerifyMode        string      `mapstructure:"ssl_verify_mode"`
+	ValidationClientName string      `mapstructure:"validation_client_name"`
+	ValidationKeyPath    string      `mapstructure:"validation_key_path"`
+	Version              string      `mapstructure:"version"`
+
+	installChefClient func(terraform.UIOutput, communicator.Communicator) error
+	createConfigFiles func(terraform.UIOutput, communicator.Communicator) error
+	runChefClient     func(terraform.UIOutput, communicator.Communicator) error
+}
+
+// ResourceProvisioner represents a generic chef provisioner
+type ResourceProvisioner struct{}
+
+// Apply executes the file provisioner
+func (r *ResourceProvisioner) Apply(
+	o terraform.UIOutput,
+	s *terraform.InstanceState,
+	c *terraform.ResourceConfig) error {
+	// Decode the raw config for this provisioner
+	p, err := r.decodeConfig(c)
+	if err != nil {
+		return err
+	}
+
+	// Set some values based on the targeted OS
+	switch s.Ephemeral.ConnInfo["type"] {
+	case "ssh", "": // The default connection type is ssh, so if the type is empty use ssh
+		p.PreventSudo = p.PreventSudo || s.Ephemeral.ConnInfo["user"] == "root"
+		p.installChefClient = p.sshInstallChefClient
+		p.createConfigFiles = p.sshCreateConfigFiles
+		p.runChefClient = p.runChefClientFunc(linuxConfDir)
+	case "winrm":
+		p.PreventSudo = true
```

Comment:
Yes... `p.UseSudo` would imply that by setting the option you are enabling all commands to use sudo. But the code is actually a little smarter (or at least it tries to be :wink:) and also looks at what user you're using and if your `root`, it will not use sudo. So `.p.PreventSudo` is a much more accurate value for this use case as it implies your OK with using sudo if needed.

Nevertheless I do agree it's not the most prettiest way of doing things, so I refactored it a little. We still have the `p.PreventSudo` option, but internally a `p.useSudo` is used. This actually makes it a little cleaner as I now no longer overwrite and "abuse" the `p.PreventSudo` configuration option for the internal logic. So a nice improvement which even made me find a small logical error in the tests :smiley: